AVX512 types + wrappers for Rust 1.89 release

Tritonn204 opened the pull request:
I took it upon myself to add AVX512 functionality to this crate for my own use, but apparently, moving AVX512 features to stable is scheduled for milestone 1.89.

That version is slated to release on the 7th, so I figured I could PR immediately and save you the trouble of catching up after the fact.

All AVX512 doctests pass in the SDE emulator.

Tritonn204 committed:
Update README.md

Tritonn204 committed:
Added AVX512 types and wrappers

Tritonn204 committed:
Activated + corrected AVX512 doctests, removing avx512_tests.rs

Lokathor commented:
So two things come to mind without looking at the code yet:

1) there's no CI changes here. Can you set up a CI pass that somehow runs the tests? Gets the emulator and builds and runs within the emulator and all that.
2) I guess we should probably set the minimum rust version field in the Cargo.toml, since this will be the first version of the crate in ages that's intended for a specific minimum rust version.

Lokathor commented:
3) this doesn't appear to implement the `core::ops` traits. To maintain consistency with the 128-bit and 256-bit types the ops traits should be implemented for the f32 and f64 carrying types, as well as implementing just the bitwise ops for the integer types (since those ops don't depend on the lane width).

Tritonn204 commented:
Sounds good, I'll get on both. Thanks for the quick initial review

Tritonn204 commented:
Minimum version would be clean, and shouldn't pose any devops issues. I'm sure anyone else who has injected AVX512 themselves already has the setup to workaround version requirements, and the standard user base for your crate already has the expectation that only stable features are enabled in the official code. 

Upgrading their Rust setup shouldn't come as an unexpected change in those cases.

Tritonn204 committed:
Add SDE emulator tests for AVX512 in CI

Lokathor commented:
Actually, I think the CI will also fail on the i586 targets once they're attempted. I think the targets were removed from rustup since the last time this repo had to do CI. If so, we can honestly just comment those targets out of CI.

Tritonn204 committed:
Altered SDE download strategy in CI

Tritonn204 committed:
Altered SDE download strategy in CI

Tritonn204 committed:
Altered SDE download strategy in CI

Tritonn204 committed:
Updated SDE URL using info from Chrome console

Tritonn204 committed:
Updated SDE exe name

Tritonn204 committed:
Updated SDE exe name

Tritonn204 committed:
Removed stale module reference from test/integration/main.rs

Tritonn204 committed:
Temporarily use only nightly when testing AVX512

Tritonn204 committed:
Corrected SDE cache strategy

Tritonn204 committed:
Corrected SDE cache strategy

Tritonn204 committed:
Corrected SDE cache strategy

Tritonn204 committed:
Corrected SDE cache strategy

Tritonn204 committed:
Corrected SDE cache strategy

Tritonn204 committed:
Corrected SDE cache strategy

Tritonn204 committed:
Final touches for CI

Tritonn204 committed:
Bumped lowest CI toolchain version

Tritonn204 commented:
> Actually, I think the CI will also fail on the i586 targets once they're attempted. I think the targets were removed from rustup since the last time this repo had to do CI. If so, we can honestly just comment those targets out of CI.

I have commented those targets out, and added+tested SDE emulation in CI for AVX512. Could you have a look and let me know what your thoughts are?

Also, the 1.51.0 (Which I changed to 1.59.0 to see if it would help) target tests fail. Should we bump that up to 1.89.0 on release? Or do you think having it be the bare minimum that does work would be better? (since the stable target passes, I imagine some target feature conventions were changed between rust 1.5x and now that are already present in some version I haven't IDed yet)

Lokathor commented:
Since this is intended for 1.89, which isn't yet released, then the minimum rust version in the Cargo.toml should actually prevent the build from working on the stable channel. Also, I'm not sure if the fixed version channel will work properly when the version isn't out yet.

For now, let's just have CI run using beta and nightly. We can adjust it later. Otherwise it all looks good.

I think for the CI to actually run against this PR you need to remove the "draft" status.

Tritonn204 commented:
> Since this is intended for 1.89, which isn't yet released, then the minimum rust version in the Cargo.toml should actually prevent the build from working on the stable channel. Also, I'm not sure if the fixed version channel will work properly when the version isn't out yet.
> 
> For now, let's just have CI run using beta and nightly. We can adjust it later. Otherwise it all looks good.
> 
> I think for the CI to actually run against this PR you need to remove the "draft" status.

Sounds good, I'll get on cross-checking the core ops then and remove draft status

Tritonn204 commented:
Do note that when CI was tested manually, stable did indeed pass, interestingly enough

Tritonn204 committed:
Added core ops + doctests for AVX512, temporary CI adjustments

Tritonn204 commented:
Converted back to draft now, as I need to make sure all intrinsics that I need for my fork of `wide` are present. Manual trigger of the current CI works great.

Lokathor commented:
I think with this addition we can call the crate a 1.0, then also update wide to go with it.

Tritonn204 committed:
added maskz wrappers and  m512i impl

Tritonn204 committed:
corrected avx512 cmp mask type suffixes

Tritonn204 committed:
Cleaned up cmp op methodology for AVX512

Tritonn204 committed:
Corrected mask op names

Tritonn204 committed:
Corrected conversion names

Tritonn204 committed:
added mask conversion methods for wide API ease-of-use

Tritonn204 committed:
added missing wrappers

Tritonn204 committed:
added keep_high multiplication for AVX512

Tritonn204 committed:
Added missing helpers

Tritonn204 committed:
Added 32-wide unpack wrappers

Tritonn204 committed:
Added m512i min/max wrappers

Tritonn204 committed:
Renamed AVX512 mask extraction methods

Tritonn204 committed:
Added 64-wide min/max

Tritonn204 committed:
Added more AVX512 conversion wrappers

Tritonn204 committed:
Added more AVX512 conversion wrappers

Tritonn204 committed:
Expanded shift wrappers for AVX512

Tritonn204 committed:
made shift_all variable for AVX512

Tritonn204 committed:
added signed shift_all wrappers for avx512

Tritonn204 committed:
added generic float lane shuffles for AVX512

Tritonn204 committed:
renamed blend_* to blend_varying_*

Tritonn204 committed:
added missing shuffle wrapper

Tritonn204 committed:
Added rounding for AVX512

Tritonn204 committed:
Added abs for AVX512

Tritonn204 committed:
Added saturating add/sub wrappers

Tritonn204 committed:
Added cmp_int_op!

Tritonn204 committed:
Corrected mm op flag list

Tritonn204 committed:
Corrected new AVX512 Doctests

Tritonn204 committed:
Make every shift_all take unsigned count param

Tritonn204 committed:
Make every shift_all take unsigned count param

Tritonn204 committed:
use epu instead of epi+cast in shifts

Tritonn204 committed:
revert prior

Tritonn204 commented:
This is in a good state now. Missing intrinsics were added, and everything still passes. Pointing my AVX512 wide fork at this allows all of those AVX512 tests to pass in SDE too, so it should be good to go.

Lokathor commented:
I'm not sure why the CI isn't running here, but it is running on your repo, https://github.com/Tritonn204/safe_arch/actions, and it looks like you've got some lingering build errors about const declarations not being imported.

Tritonn204 commented:
Ah, didn't notice `avx512er` is yet another AVX512 feature separation module. This will be quick to fix + re-validate

Tritonn204 committed:
corrected AVX512 round gating + doctest const qualifiers

Tritonn204 committed:
Removed unrecognized low-level feature flag from workflow

Tritonn204 commented:
[Fixed](https://github.com/Tritonn204/safe_arch/actions/runs/16816697356)

Lokathor commented:
Alright I'll merge now, but will hold off on a publish until 1.89 comes out.

Lokathor commented:
Oh wait 1.89 is today! I'll see about a publish some time soon then.

Lokathor commented:
Released as 0.9, if there's no problems we can move this version into being a 1.0 version at some point.
